### PR TITLE
[SYCL-MLIR] Allow accessing vector elements using array subscripts

### PR DESCRIPTION
**Pull request by victor-eds**
Use vector.extractelement to obtain elements of a vector.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -870,6 +870,26 @@ ValueCategory MLIRScanner::VisitConstructCommon(clang::CXXConstructExpr *Cons,
...
+    Expr->dump();
+    llvm::dbgs() << "\n";
+  });
+
+  assert(!Expr->getBase()->getType()->isVLSTBuiltinType() &&
+         "Not supported yet");
+
+  if (Expr->getBase()->getType()->isVectorType()) {
+    // Handle the vector case.
+    ValueCategory Base{Visit(Expr->getBase()).getValue(Builder), false};
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:884]: This line should be replaced with:

```cpp
    auto Base = Visit(Expr->getBase());
```

when we implement obtaining lvalues from expressions as this one.

**Review by victor-eds — COMMENTED** [review]
Note that this only allows obtaining an element of a vector as an rvalue. Obtaining an lvalue from a vector is currently on the pipeline.

This is needed to support `sycl::vec`.

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -870,6 +870,26 @@ ValueCategory MLIRScanner::VisitConstructCommon(clang::CXXConstructExpr *Cons,
...
+    auto Idx = Visit(Expr->getIdx());
+
+    mlirclang::warning() << "Not emitting bounds check\n";
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:887]: I'm wondering whether we really need this warning. Is the subscript expression is out of bounds the program is incorrect.  I do not think we should generate any code to check that the index expression is within the declared array bounds, with that in mind I think we can remove this kind of warnings.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 3]
```diff
@@ -870,6 +870,26 @@ ValueCategory MLIRScanner::VisitConstructCommon(clang::CXXConstructExpr *Cons,
...
+  LLVM_DEBUG({
+    llvm::dbgs() << "VisitArraySubscriptExpr: ";
+    Expr->dump();
+    llvm::dbgs() << "\n";
+  });
+
+  assert(!Expr->getBase()->getType()->isVLSTBuiltinType() &&
+         "Not supported yet");
+
+  if (Expr->getBase()->getType()->isVectorType()) {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:882]: Would it make sense to have a set of private member functions to deal with the case where `Expr` has vector type. This function could delegate code generation to those functions when the expr is a vector.

**Review by etiotto — APPROVED** [review]
Left some inline suggestions (can be implemented separately).

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 3 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:950]: Done

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 2 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc]: If `clang` is passed `-fsanitize=bounds`, the code generated actually differs, checking that the index is inbounds (for variable indices). That's why I'm adding this kind of warnings. As we might want to add this functionality in the future, I'd say we want to keep this for now. Do you agree?

Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGExpr.cc]: Yes we can keep it for now, in a future PR we should then emit this warning only when `-fsanitize=bounds` is set.

**Pull request merged**